src/app/core/services/notifications/fcm-rest-notification.service.ts

```diff
@@ -0,0 +1,186 @@
+import { LocalizationService } from '../misc/localization.service'
+import { LogService } from '../misc/log.service'
+import { ScheduleService } from '../schedule/schedule.service'
+import { StorageService } from '../storage/storage.service'
+import { FcmNotificationService } from './fcm-notification.service'
+import { NotificationGeneratorService } from './notification-generator.service'
+
+@Injectable()
+export class FcmRestNotificationService extends FcmNotificationService {
+  private APP_SERVER_URL = 'http://localhost:8080'
+  apiClient
+
+  constructor(
+    public notifications: NotificationGeneratorService,
+    public storage: StorageService,
+    public schedule: ScheduleService,
+    public config: SubjectConfigService,
+    public firebase: Firebase,
+    public platform: Platform,
+    public logger: LogService,
+    public remoteConfig: RemoteConfigService,
+    public localization: LocalizationService
+  ) {
+    super(storage, config, firebase, platform, logger, remoteConfig)
+    this.initApiClient()
+  }
+
+  async initApiClient() {
+    await Swagger({ url: `${this.APP_SERVER_URL}/v3/api-docs` }).then(
+      client => {
+        this.apiClient = client
+      }
+    )
+  }
+
+  getSubjectDetails() {
+    return this.checkProjectAndSubjectExistElseCreate()
+  }
+
+  publishAllNotifications(user, sourceId, limit): Promise<any> {
+    return this.schedule.getTasks(TaskType.ALL).then(tasks => {
+      const fcmNotifications = this.notifications
+        .futureNotifications(tasks, limit)
+        .map(t => this.format(t, sourceId))
+      this.logger.log('NOTIFICATIONS Scheduling FCM notifications')
+      this.logger.log(fcmNotifications)
+      return Promise.all(
+        fcmNotifications
+          .map(n => this.sendNotification(n, user.subjectId, user.projectId))
+          .concat([this.setLastNotificationUpdate()])
+      )
+    })
+  }
+
+  publishTestNotification(user, sourceId): Promise<void> {
+    return this.sendNotification(
+      this.format(this.notifications.createTestNotification(), sourceId),
+      user.subjectId,
+      user.projectId
+    )
+  }
+
+  cancelAllNotifications(user): Promise<any> {
+    return this.apiClient.apis[
+      'fcm-notification-controller'
+    ].deleteNotificationsForUser({
+      subjectId: user.subjectId,
+      projectId: user.projectId
+    })
+  }
+
+  cancelSingleNotification(user, notificationId) {
+    return this.apiClient.apis['fcm-notification-controller']
+      .deleteNotificationUsingProjectIdAndSubjectIdAndNotificationId({
+        subjectId: user.subjectId,
+        projectId: user.projectId,
+        id: notificationId
+      })
+      .then(() =>
+        console.log('Success cancelling notification ' + notificationId)
+      )
+  }
+
+  private checkProjectAndSubjectExistElseCreate(): Promise<any> {
+    return this.checkProjectExistsElseCreate().then(() =>
+      this.checkSubjectExistsElseCreate()
+    )
+  }
```

Comment:
These functions should be extracted to a general AppServerService since these are not directly linked with scheduling of notifications. More over, these need to be called on start of the application, to create a project and a subject if they don't exist, even if there are no notifications to schedule. This will also contain functions to update the timezone of the user, in case it has changed.

Reply:
Ok I've added an `AppServerService`